ddp-workspace/projects/ddp-dsm-ui/src/app/FON/components/paginator/paginator.component.ts

```diff
@@ -0,0 +1,143 @@
+  from: number;
+  to: number;
+}
+
+enum SHIFT_PAGE {
+  START = 'START',
+  END = 'END',
+  ONE_MORE = 'ONE_MORE',
+  ONE_LESS = 'ONE_LESS'
+}
+
+/* Defaults */
+const PAGE_SIZE_OPTIONS = [10, 25, 50];
+const ROWS_PER_PAGE = 10;
+const CURRENT_PAGE_INDEX = 1;
+const VISIBLE_PAGES = 3;
+
+
+@Component({
+  selector: 'app-paginator',
+  templateUrl: './paginator.component.html',
+  styleUrls: ['./paginator.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class PaginatorComponent implements OnChanges {
+  @Input() totalCount: number;
+  @Input() rowsPerPage = 10;
+  @Input() currentPageIndex: number;
+  @Input() pageSizeOptions: number[];
+  @Input() visiblePages: number;
+
+  @Output() pageChanged: EventEmitter<pageProps> = new EventEmitter();
+
+  constructor() {}
+
+  ngOnChanges(changes: SimpleChanges): void {
+    this.setDefaultParams();
+    this.generatePagesArray();
+  }
+
+  public get currentPageRange(): (string|number)[] {
+    const currentRange = this.filteredCurrentPageRange;
+    this.fillWithDots(currentRange);
+    return currentRange;
+  }
+
+  public changePage(page: number): void {
+    if (typeof page === 'number') {
+      this.currentPageIndex = page;
+      this.pageChanged.emit(this.pageProps);
+    }
+  }
+
+  public shiftPage(order: string): void {
+    switch (order) {
+      case SHIFT_PAGE.START:
+        this.currentPageIndex = 1;
+        break;
+      case SHIFT_PAGE.END:
+        this.currentPageIndex = this.generatePagesArray().length;
+        break;
+      case SHIFT_PAGE.ONE_MORE:
+        this.currentPageIndex = Math.min(this.currentPageIndex + 1, this.generatePagesArray().length);
+        break;
+      case SHIFT_PAGE.ONE_LESS:
+        this.currentPageIndex = Math.max(this.currentPageIndex - 1, 1);
+        break;
+      default:
+        return;
+    }
+    this.pageChanged.emit(this.pageProps);
+  }
+
+  public setRows(rows: number): void {
+    this.rowsPerPage = rows;
+    this.changePage(1);
+    this.generatePagesArray();
+  }
+
+  /* Paginator Engine */
+
+  private setDefaultParams(): void {
+    if(!this.rowsPerPage || typeof this.rowsPerPage !== 'number')
+      {this.rowsPerPage = ROWS_PER_PAGE;}
+    if(!(this.pageSizeOptions instanceof Array) || this.pageSizeOptions.length < 1)
+      {this.pageSizeOptions = PAGE_SIZE_OPTIONS;}
+    if(!this.currentPageIndex || typeof this.currentPageIndex !== 'number')
+      {this.currentPageIndex = CURRENT_PAGE_INDEX;}
+    if(!this.visiblePages || typeof this.visiblePages !== 'number')
+      {this.visiblePages = VISIBLE_PAGES;}
+  }
+
+  private get filteredCurrentPageRange(): any[] {
+    return this.generatePagesArray()
+      .filter(
+        (page, _, array) => this.displayOrNot(page, array)
+      );
+  }
+
+  private fillWithDots(pagesArray: any[]): void {
+    pagesArray
+      .forEach(
+        (page, index, array) =>
+          page - array[index + 1] < -1 &&
+          array.splice(index + 1, 0, '...')
```

Comment:
Generally, it is not a good idea to mutate an array inside of `forEach` loop.

Reply:
In this case, we are not mutating the original array, but the one we have kept in currentRange variable (I suppose, I have compared instances and they are different). Btw, we need to insert dots somehow, that's why I decided to perform that action.

Reply:
In this scenario: 
![image](https://user-images.githubusercontent.com/77500504/177133466-f17202d5-7355-44c2-9ce9-0a0e6b8a663d.png)

Reply:
I mean the input array (pagesArray) does mutate during iteration through itself:
```
   pagesArray
      .forEach(
        (page, index, array) =>
          page - array[index + 1] < -1 &&
          array.splice(index + 1, 0, '...')
      );
```
it is possible in that case because JS is a quite flexible language and does lots of unexpectable things(e.g. can compare a sting and a number, etc.). 
 But generally it 's better to avoid such stuff ([an issue example](https://javascript.plainenglish.io/why-the-foreach-js-method-might-not-always-be-the-best-choice-5eeaf048c3a2#:~:text=Some%20other%20cases,the%20calling%20array.)) 
 Suggest using for example a `reduce` operator for creating a new transformed array.
 
 You are not required to do it right now, just to keep in mind.

Reply:
Alright. Got it now. Thanks, I will do this as well, with other changes.

Reply:
Removed!